### PR TITLE
[core] add bucket padding to tpu_model_runner

### LINKED ISSUE #14581: [Feature][Hardware][TPU]: Improve the token_num padding logic
### 🚀 The feature, motivation and pitch

Currently the token_num is padded to power of 2. It is quite a waste of computation when token_num is large. In the meantime, in one of our benchmarking, the best max-num-batched-tokens is 512, also according to https://jax-ml.github.io/scaling-book/roofline/, we don't really need max-num-batched-tokens to be very large.

Also cudagraph precompiles for [512,504,496,488,480,472,464,456,448,440,432,424,416,408,400,392,384,376,368,360,352,344,336,328,320,312,304,296,288,280,272,264,256,248,240,232,224,216,208,200,192,184,176,168,160,152,144,136,128,120,112,104,96,88,80,72,64,56,48,40,32,24,16,8,4,2,1], we can use a similar padding strategy for TPU.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Chenyaaang**
Add bucket padding to tpu, instead of padding to the power of 2, if num_token < bucket_padding_gap, pad to the nearest power of 2, if num_token > bucket_padding_gap, the padding size is increased by bucket_padding_gap. 
For example, bucket_padding_gap = 64, max_num_batch_tokens = 512, then the paddings will be 16, 32, 64, 128, 192, 256, 320, 384, 448, 512. This helps reduce the computation cost for large num_tokens, e.g. num_tokens = 300, instead of padding to 512, now pad to 320. 

FIX #14581

### CODE REVIEW

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -158,6 +159,23 @@ def __init__(
         # Range tensor with values [0 .. self.max_num_tokens - 1].
         # Used to initialize positions / context_lens / seq_lens
         self.arange_np = np.arange(self.max_num_tokens, dtype=np.int32)
+        self.paddings = self._get_paddings()
+
+    def _get_paddings(self) -> list[int]:
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:164]: It will be great if we can have unit tests for these functions which only exist in TPU platform, cc @yarongmu-google

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -809,6 +823,13 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
             self.vllm_config.compilation_config.static_forward_context,
             self.kv_caches)
 
+    def _get_padded_token_len(self, x: int) -> int:
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:826]: ditto

**Comment by yaochengji** [comment]
@Chenyaaang Thanks for your contribution, left some comments above!

@robertgshaw2-redhat I know there's another configuration option `cudagraph_capture_sizes`, which is similar to `bucket_padding_gap` in this PR, do you think we should merge them into one, and has default value for different platform?

**Review by alexm-redhat — APPROVED** [review]
Good idea!

**Comment by DarkLight1337** [comment]
Please fix the merge conflict

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -158,6 +160,9 @@ def __init__(
         # Range tensor with values [0 .. self.max_num_tokens - 1].
         # Used to initialize positions / context_lens / seq_lens
         self.arange_np = np.arange(self.max_num_tokens, dtype=np.int32)
+        self.paddings = _get_paddings(
```
Comment (NickLucche) [inline vllm/v1/worker/tpu_model_runner.py:163]: nit: is `num_tokens_paddings` too verbose of a name?

**Review by NickLucche — CHANGES_REQUESTED** [review]
looks good! Also in favor of unifying with `cudagraph_capture_sizes`

**Review by yaochengji — APPROVED** [review]
LGTM, thanks!

File: vllm/v1/worker/tpu_model_runner.py [hunk 4]
```diff
@@ -928,12 +925,32 @@ def _get_padded_number(n: int, multiple: int) -> int:
...
-        return 16
-    return 1 << (x - 1).bit_length()
-
-
 def _get_padded_num_reqs_with_upper_limit(x, upper_limit) -> int:
     res = MIN_NUM_SEQS if x <= MIN_NUM_SEQS else 1 << (x - 1).bit_length()
     return min(res, upper_limit)
+
+
+def _get_paddings(min_token_size, max_token_size, padding_gap) -> list[int]:
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/tpu_model_runner.py:933]: type hints are needed

File: vllm/v1/worker/tpu_model_runner.py [hunk 5]
```diff
@@ -574,7 +579,7 @@ def execute_model(
 
         # Prepare inputs
         attn_metadata, logits_indices = self._prepare_inputs(scheduler_output)
-
+        print("logits_indices: ", logits_indices)
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/tpu_model_runner.py:582]: nit

File: vllm/engine/arg_utils.py [hunk 6]
```diff
@@ -1326,8 +1334,7 @@ def create_engine_config(
             scheduler_cls=self.scheduler_cls,
             max_num_partial_prefills=self.max_num_partial_prefills,
             max_long_partial_prefills=self.max_long_partial_prefills,
-            long_prefill_token_threshold=self.long_prefill_token_threshold,
-        )
+            long_prefill_token_threshold=self.long_prefill_token_threshold)
```
Comment (robertgshaw2-redhat) [inline vllm/engine/arg_utils.py:1337]: nit

File: vllm/engine/arg_utils.py [hunk 7]
```diff
@@ -130,6 +130,7 @@ class EngineArgs:
     long_prefill_token_threshold: Optional[int] = 0
     max_num_seqs: Optional[int] = None
     max_logprobs: int = 20  # Default value for OpenAI Chat Completions API
+    tpu_bucket_padding_gap: Optional[int] = None
```
Comment (robertgshaw2-redhat) [inline vllm/engine/arg_utils.py:133]: I do not think this `engine_arg` is being propogated anywhere?

**Review by robertgshaw2-redhat — COMMENTED** [review]
Do you think that we should allow the user to specify this? I would think that we should try to keep this internal unless there is a strong reason why someone would toggle this?

**Review by robertgshaw2-redhat — CHANGES_REQUESTED** [review]
blocking merge until questions addressed

File: vllm/engine/arg_utils.py (hunk 7 above)
Comment (Chenyaaang) [reply vllm/engine/arg_utils.py:133]: It's in compilation config: https://github.com/vllm-project/vllm/blob/main/vllm/engine/arg_utils.py#L232

Comment (robertgshaw2-redhat) [reply vllm/engine/arg_utils.py:133]: I know, but there is no code for updating this value based on the engine arg

File: vllm/config.py [hunk 8]
```diff
@@ -3034,6 +3034,12 @@ class CompilationConfig(BaseModel):
     cudagraph_capture_sizes: Optional[list[int]] = None
     cudagraph_copy_inputs: bool = False
 
+    # TODO: combine tpu and cuda padding together.
+    # The gap between each bucket for padding (in tpu).
+    # For example, the bucket padding is (16, 24, 32, 48, ..., 512),
```
Comment (vanbasten23) [inline vllm/config.py:3039]: nit: missing a 40 in `16, 24, 32, 48`?

File: vllm/engine/arg_utils.py (hunk 7 above)
Comment (Chenyaaang) [reply vllm/engine/arg_utils.py:133]: oh yes, thanks for pointing this out.

File: vllm/config.py (hunk 8 above)
Comment (Chenyaaang) [reply vllm/config.py:3041]: fixed, thanks

**Comment by Chenyaaang** [comment]
> Do you think that we should allow the user to specify this? I would think that we should try to keep this internal unless there is a strong reason why someone would toggle this?

I think the param is useful for online serving, by exposing this parameter, users can adjust the gap based on the model size. My understanding is for example the gap is small then it will lead to a long pre-compile time, that's something customer can adjust.

**Comment by robertgshaw2-redhat** [comment]
This PR seems to have broken CUDA

**Comment by alexm-redhat** [comment]
@Chenyaaang thanks for reducing the padding gap, this is useful. Could you please address @robertgshaw2-redhat comment and fix the build (so we can merge it). Thanks!

**Comment by Chenyaaang** [comment]
> @Chenyaaang thanks for reducing the padding gap, this is useful. Could you please address @robertgshaw2-redhat comment and fix the build (so we can merge it). Thanks!

I've fixed the build and replied to @robertgshaw2-redhat's comment.

File: tests/tpu/test_compilation.py [hunk 9]
```diff
@@ -52,16 +52,15 @@
...
-# 4. forward pass (shape 32)
+# 2. forward pass (shape 16)
+# 3. forward pass (shape 32)
 # and later calls should not trigger Dynamo compilation again.
 # NOTE: It might still trigger XLA compilation.
 
-# Check we have 4 compiled codes
-assert len(compiled_codes) == 4
+# Check we have 3 compiled codes
+assert len(compiled_codes) == 3
```
Comment (lsy323) [inline tests/tpu/test_compilation.py:63]: @Chenyaaang Thanks for looking into this! Could we branch out for v0 and v1? In v0 it should be 4 compiled code.

Comment (Chenyaaang) [reply tests/tpu/test_compilation.py]: Done, thanks!

**Review by robertgshaw2-redhat — APPROVED** [review]
